bpf: Re-introduce ICMPv6 NS responder on from-netdev

jschwinger233 opened the pull request:
This reverts commit 658071414ca4606e537bc4bbb37dcae5e18cd7dc, to fix the breakage of "IPv6 NS responder for pod" introduced by https://github.com/cilium/cilium/pull/12086 (bpf: Reply NA when recv ND for local IPv6 endpoints).

658071414ca4606e537bc4bbb37dcae5e18cd7dc was merged to solve https://github.com/cilium/cilium/issues/14509. To not revive #14509, this commit also passes through ICMPv6 NS if the target is native node IP (eth0's addr). By letting stack take care of those NS-for-node-IP packets, we managed to:

1. Solve #14509 again, but in a way keeping NS responder. The cause of #14509 was NS responder always generates ND whose source IP is "router_ip" (cilium_internal_ip) rather than "node_ip". Once we pass those NS-for-node-IP packets to stack, the ND response would naturally have "node_ip" as source.

2. Avoid the fib_lookup failure mentioned at https://github.com/cilium/cilium/pull/30837#issuecomment-1960897445.

This PR also adds bpf unit test to cover IPv6 NS responder feature.

Fixes: #30926 

```release-note
Fixes an IPv6 issue that cilium doesn't respond to Neighbor Solicitation targeting the pods on same node.
```

jschwinger233 committed:
bpf: Re-introduce ICMPv6 NS responder on from-netdev

This reverts commit 658071414ca4606e537bc4bbb37dcae5e18cd7dc, to fix
the breakage of "IPv6 NS responder for pod" introduced by
https://github.com/cilium/cilium/pull/12086 (bpf: Reply NA when recv ND for local IPv6 endpoints).

658071414ca4606e537bc4bbb37dcae5e18cd7dc was merged to solve https://github.com/cilium/cilium/issues/14509.
To not revive #14509, this commit also passes through ICMPv6 NS if the
target is native node IP (eth0's addr). By letting stack take care of
those NS-for-node-IP packets, we managed to:

1. Solve #14509 again, but in a way keeping NS responder. The cause of
   #14509 was NS responder always generates ND whose source IP is
   "router_ip" (cilium_internal_ip) rather than "node_ip". Once we pass
   those NS-for-node-IP packets to stack, the ND response would
   naturally have "node_ip" as source.

2. Avoid the fib_lookup failure mentioned at https://github.com/cilium/cilium/pull/30837#issuecomment-1960897445.

icmp6_host_handle() also has a new parameter `handle_ns` to control if
we want NS responder to be active. If it is called from `to-netdev` code
path, handle_ns is set to false. This is suggested by julianwiedmann.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
bpf/tests: Add IPv6 NDP bpf test

This commit adds bpf/tests/ipv6_ndp_from_netdev_test.c to cover two
scenarios:
1. from_netdev receives IPv6 NS for a pod IP on the same host
2. from_netdev receives IPv6 NS for the node IP (eth0's addr)

For case 1, from_netdev should return a NA on behalf of the target pod
to avoid https://github.com/cilium/cilium/issues/30926.
for case 2, it must return the NS to stack to address https://github.com/cilium/cilium/issues/14509.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

julianwiedmann commented on bpf/lib/icmp6.h:
```diff
@@ -309,15 +310,27 @@ static __always_inline int __icmp6_handle_ns(struct __ctx_buff *ctx, int nh_off)
+			 * source IP, to avoid bpf_fib_lookup failure as mentioned at
+			 * https://github.com/cilium/cilium/pull/30837#issuecomment-1960897445.
+			 */
+			return CTX_ACT_OK;
```
Just to note - as this is executed in a tail-call, such traffic then skips all subsequent code that would normally be present in `from-netdev`. That's quite subtle :slightly_smiling_face:.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -495,7 +494,7 @@ handle_to_netdev_ipv6(struct __ctx_buff *ctx, struct trace_ctx *trace, __s8 *ext
 
 	if (likely(nexthdr == IPPROTO_ICMPV6)) {
-		ret = icmp6_host_handle(ctx, ETH_HLEN + hdrlen);
+		ret = icmp6_host_handle(ctx, ETH_HLEN + hdrlen, ext_err);
```
This user would now also pick up the changed behaviour in `icmp6_host_handle()`. Is that intentional, or should we pass a flag that we don't want the NS responder to be active for this code path?

julianwiedmann commented on bpf/lib/icmp6.h:
```diff
@@ -446,9 +463,6 @@ icmp6_host_handle(struct __ctx_buff *ctx, int l4_off)
 
-	if (type == ICMP6_NS_MSG_TYPE)
-		return CTX_ACT_OK;
-
```
Do we want to keep this now, to tolerate NS for `!handle_ns` ?

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -144,31 +144,28 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx __maybe_unused,
 		return DROP_INVALID;
 
-#ifdef ENABLE_HOST_FIREWALL
 	nexthdr = ip6->nexthdr;
 	hdrlen = ipv6_hdrlen(ctx, &nexthdr);
 	if (hdrlen < 0)
 		return hdrlen;
 
```
Looking at it, I think we could limit this code section to `is_defined(ENABLE_HOST_FIREWALL) || !from_host`. But let's consider that in a follow-up.

julianwiedmann reviewed (approved):
lgtm, thank you Gray! Let's hope that CI still likes it :slightly_smiling_face:.

jschwinger233 committed:
bpf/tests: Remove SKIP_ICMPV6_NS_HANDLING from tc_nodeport_l3_dev.c

SKIP_ICMPV6_NS_HANDLING was there to pass bpf coverage test, which is
gone by https://github.com/cilium/cilium/pull/28090.

In the meantime, removing SKIP_ICMPV6_NS_HANDLING from
tc_nodeport_l3_dev.c prevents "potential missed tailcall" errors
introduced by https://github.com/cilium/cilium/pull/30467, as
tail_icmp6_handle_ns() doesn't exist when SKIP_ICMPV6_NS_HANDLING is
defined, but still gets tail-called by icmp6_handle_ns().

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>
